### PR TITLE
Separate features entirely from GLT Dataset

### PR DESCRIPTION
**Pull request by Elnifio**
This PR separates the features entirely from GLT Dataset object. From our local testing, this shows 2x speedup on 3-layer models compared to not pinning the features on GLT Dataset object. 

The following data are gathered on hyperparameter (all other remains default): 
- Dataset: in-memory FP16 IGBH-Small
- GPU: 1x2 A6000
- Local batch size: 512 
- Fanout: 15,10,5; Number of layers: 3
- Learning rate: 0.001
- Epochs: 3

When using vanilla GLT training **with** `pin_feature`, the per-epoch time is 4min 44s, total time 14min 6s: 
image

When using **this workaround**, the per-epoch time is 5min 55s, total time 17min 42s:
image

When using GLT training **without** `pin_feature`, the per-epoch time is 10min 52s, total time 32min 33s: 
image

### CODE REVIEW

**Comment by Elnifio** [comment]
We see that this effect is also reproduceable on 2-layer models. 

The following data are gathered on hyperparameter (all other remains default):
- Dataset: in-memory FP16 IGBH-Small
- GPU: 1x2 A6000
- Local batch size: 512
- **Fanout: 15,10; Number of layers: 2**
- Learning rate: 0.001
- Epochs: 3

When using GLT with `pin_feature`, the per-epoch time is 50s, total time is 2min 26s: 
image

When using this workaround, the per-epoch time is 1min 7s, total time is 3min 19s:
image

When using GLT without `pin_feature`, the per-epoch time is 2min 1s, total time 6min 1s: 
image

**Comment by LiSu** [comment]
The reason for these results is that when using FP16, the GLT script converts the FP16 feature in the batch results into FP32 before the feature is transferred to the training device. This issue is addressed in #105 by explicitly copying the feature to the GPU before converting the data type.
